### ISSUE TITLE
[Feature Request] Delete Email preview or convert HTML entities used in Emails preview hacks

### ISSUE THREAD

**Issue by tkapias** [jugglerchris/rust-html2text#134]
## Introduction

In Email, it was a common practice to add a separator indicating the end of the preview text.

This method does not work on all clients and is subject to regular changes. I came across [this](https://www.litmus.com/blog/the-little-known-preview-text-hack-you-may-want-to-use-in-every-email) article on the subject with a few examples.

## Issue

The problem is that the invisible characters used for this hack are badly converted by html2text, which leaves many ["COMBINING GRAPHEME JOINER"](https://www.compart.com/en/unicode/U+034F) entities in the output.

This entity is crudely displayed in pagers or TUI mail clients (dotted circle in a dotted square).

The first complication is that there is many variants of the sequences used, some are presented in the article above. In my own inbox, this week, I found at least 2 variants:

- a repetition of `&#847;&zwnj;&nbsp;`
- a repetition of `­͏ ‌ &nbsp`, which you can reproduce with `echo -ne '\u00ad\u034f \u200c &nbsp;'`

And another complication is that the sequence is sometime formated in columns with newlines at different places in the sequence, like here:

![image](https://github.com/jugglerchris/rust-html2text/assets/45816387/4a250527-e4b8-4aa9-9547-1ca726c07fad)

This exemple would currently be converted by html2text like that:

![image](https://github.com/jugglerchris/rust-html2text/assets/45816387/9fe3ab24-9004-4e68-a98b-94c9037b82df)

## Solution

For now I was piping the source file to sed first to delete those sequences:

`sed -e 's/&#847;&zwnj;&nbsp;/\n/g' -e "s/$(echo -ne '\u00ad\u034f')[; ^C]*$(echo -ne '\u200c')[; \n]*&nbsp[; \n]*/\n/g"`

I updated this script 3 times and I was going to find a solution for the cases with columns/newlines.

But I've just realized that the best solution would be to get rid of all the text before this sequence in the `body`, since it's a preview, a repetition of the text present in the rest of the document.

I think that such a deletion would impact only email documents and be a good addition to the library even by default.

**Comment by jugglerchris** [comment]
I've never come across this hack, and I have used html2text in my TUI email client from the start!

I would not be happy matching some random sequences and deleting content before them, but I'd be open to some kind of API to make it possible to implement this kind of thing if you have any ideas.

**Comment by tkapias** [comment]
These preview headers are mostly present in newsletters or ads, so you may be dismissing this them as spam, or they may not be using the version with the `u+034f` and be less noticeable.

I've found a lot of sites that recommend this kindod addition by searching for `email html preview preheader` or `email html preview hack`.
But as I read these tutorials, **I realized that most of these inserts are enclosed in a `div` set to `display: none`. I hadn't realized this before.**

Then the question becomes "why does html2text display these hidden elements?"
And if there's no good reason, **maybe we should just remove them from the rendering.**

**Comment by tkapias** [comment]
An example of open source Html Email template with a hidden preview element:

`curl https://raw.githubusercontent.com/TedGoas/Cerberus/main/cerberus-fluid.html | rust-html2text`

The first paragraph sould be hidden, but html2text displays it:

```
(Optional) This text will appear in the inbox preview, but not the email body.
It can be used to supplement the email subject line or even summarize the
email's contents. Extended text preheaders (~490 characters) seems like a better
UX for anyone using a screenreader or voice-command apps like Siri to dictate
the contents of an email. If this text is not included, email clients will
automatically populate it using the text (including image alt text) at the start
of the email's body.
‌‌‌‌‌‌‌‌‌‌‌‌‌‌‌‌‌‌

[alt_text]
[alt_text]
# Praesent laoreet malesuada cursus.

Maecenas sed ante pellentesque, posuere leo id, eleifend dolor. Class aptent
taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos.
Praesent laoreet malesuada cursus. Maecenas scelerisque congue eros eu posuere.
Praesent in felis ut velit pretium lobortis rhoncus ut erat.
[Primary Button][1]
## Praesent in felis ut velit pretium lobortis rhoncus ut erat.

* A list item.
* Another list item here.
* Everyone gets a list item, list items for everyone!

Maecenas sed ante pellentesque, posuere leo id, eleifend dolor. Class aptent
taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos.
Praesent laoreet malesuada cursus. Maecenas scelerisque congue eros eu posuere.
Praesent in felis ut velit pretium lobortis rhoncus ut erat.
[alt_text]
Maecenas sed ante pellentesque, posuere leo id, eleifend dolor.
[alt_text]
Maecenas sed ante pellentesque, posuere leo id, eleifend dolor.
Maecenas sed ante pellentesque, posuere leo id, eleifend dolor. Class aptent
taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos.
Praesent laoreet malesuada cursus. Maecenas scelerisque congue eros eu posuere.
Praesent in felis ut velit pretium lobortis rhoncus ut erat.

View as a Web Page

Company Name
123 Fake Street, SpringField, OR, 97477 US
(123) 456-7890

unsubscribe

Maecenas sed ante pellentesque, posuere leo id, eleifend dolor. Class aptent
taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos.
Praesent laoreet malesuada cursus. Maecenas scelerisque congue eros eu posuere.
Praesent in felis ut velit pretium lobortis rhoncus ut erat.

[1]: https://google.com/
```

**Comment by jugglerchris** [comment]
> Then, the question becomes "why does html2text display these hidden elements?"
And if there's no good reason, maybe we should just remove them from the rendering.

I had a look at that example HTML, although the weird characters are in a `<div style="display:none">` (and so will be hidden when document CSS is enabled), the preview stuff is in a `<div style="max-height:0; overflow:hidden">`.

I think it would be reasonable for html2text to treat `max-height: 0` as an alias for `display: none`.

**Comment by jugglerchris** [comment]
Does #135 work for you?  (CSS needs to be enabled)

**Comment by jugglerchris** [comment]
I've now released it (tweaked again) as 0.12.4 - let me know if that doesn't work.

**Comment by tkapias** [comment]
Sorry for the delay.

Thank you for the update, I did not use the CSS feature before and it solves the issue.

But using CSS creates a new issue for me by overriding RichAnnotations colours, which is not adapted for a TUI email client display. I will open another issue about that.
